In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
df = pd.read_csv('hotel_booking.csv')

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 36 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [5]:
df.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  customer_type  \
0                        0                     0       2  ...      Transient   
1                        0                     0       2  ...      Transient   
2                        0                     1       1  ...      Transient   
3                        0                     1       1  ...      Transient   
4                        0                     2       2  ...      Transient   

    adr required_car_parking_spaces total_of_special_requests  \
0   0.0                           0                         0   
1   0.0                           0                         0   
2  75.0                           0                         0   
3  75.0                           0                         0   
4  98.0                           0                         1   

  reservation_status reservation_status_date            name  \
0          Check-Out              2015-07-01   Ernest Barnes   
1          Check-Out              2015-07-01    Andrea Baker   
2          Check-Out              2015-07-02  Rebecca Parker   
3          Check-Out              2015-07-02    Laura Murray   
4          Check-Out              2015-07-03     Linda Hines   

                         email  phone-number       credit_card  
0  Ernest.Barnes31@outlook.com  669-792-1661  ************4322  
1       Andrea_Baker94@aol.com  858-637-6955  ************9157  
2   Rebecca_Parker@comcast.net  652-885-2745  ************3734  
3            Laura_M@gmail.com  364-656-8427  ************5677  
4           LHines@verizon.com  713-226-5883  ************5498  

[5 rows x 36 columns]

# Stage 2

# Data Preprocessing

[Stage 2] Final Project DS 23

Periode : 15 - 21 Agustus

Halo teman-teman! Setelah menemukan insight, dan visualiasi dari dataset yang dipilih, kini saatnya teman-teman masuk ke dalam proses Data Pre-Processing untuk menentukan :
1. Apakah sudah melakukan pengecekkan data bermasalah seperti missing values, invalid values, atau data duplicate dan sudah membersihkannya?
2. Apakah sudah menentukan feature apa saja yang akan digunakan, atau perlu ditambahkan, dan reformatting feature sesuai dengan kebutuhan?

Penting! Saat memanipulasi data, mohon dilakukan dengan alasan yang jelas, dan tidak melakukan penambahan feature baru tapi tidak ada alasan yang mendasari langkah tersebut diperlukan.

Silahkan selesaikan dan kumpulkan dokumen Stage 2 tersebut (Notulen Mentoring, Source Code, Laporan Project) ke link gdrive yang sudah disiapkan (akan diinfokan secara personal ke masing-masing tim), paling lambat tanggal 21 Agustus 2022 pukul 23:59

**Ari, Emir, Diah**

>1. arrival_date_week_number >> hapus
2. adults
3. is_repeated_guest
4. previous_cancellations >> 1 atau 0
5. previous_bookings_not_canceled >> hapus
6. reserved_room_type >> fitur baru perbandingan kamar
7. assigned_room_type >> fitur baru perbandingan kamar
8. agent >> hapus
9. company >> hapus
10. reservation_status >> hapus
11. reservation_status_date >> hapus
12. name >> hapus
13. email >> hapus 
14. phone-number >> hapus
15. credit_card >> hapus

Date

In [6]:
def month(x):
    if x['arrival_date_month'] == 'January':
            month = '1'
    elif x['arrival_date_month'] == 'February':
            month = '2'
    elif x['arrival_date_month'] == 'March':
            month = '3' 
    elif x['arrival_date_month'] =='April':
            month = '4'
    elif x['arrival_date_month'] == 'May':
            month = '5'    
    elif x['arrival_date_month'] == 'June':
            month = '6'
    elif x['arrival_date_month'] == 'July':
            month = '7'
    elif x['arrival_date_month'] == 'August':
            month = '8'
    elif x['arrival_date_month'] == 'September':
            month = '9'
    elif x['arrival_date_month'] == 'October':
            month = '10'
    elif x['arrival_date_month'] == 'November':
            month = '11'
    else:
            month = '12'
    return month

In [30]:
df['month_number'] = df.apply(lambda x: month(x), axis=1)
# merubah tahun dan hari menjadi string
df["arrival_date_day_of_month"]= df["arrival_date_day_of_month"].astype(str)
df["arrival_date_year"]= df["arrival_date_year"].astype(str)
# membuat kolom date of arrival
df["Date"] = pd.to_datetime(df["arrival_date_year"] + "/" + df["month_number"] + "/" + df["arrival_date_day_of_month"])  
# menambahkan info tanggal weekend
df["isWeekend"] = df["Date"].dt.weekday >= 5
def weekend(x):
    if x['isWeekend'] ==True:
            weekend = '1'
    else: 
            weekend = '0'
    return weekend
df['is_weekend'] = df.apply(lambda x: weekend(x), axis=1)
df

hotel  is_canceled  lead_time arrival_date_year  \
0       Resort Hotel            0        342              2015   
1       Resort Hotel            0        737              2015   
2       Resort Hotel            0          7              2015   
3       Resort Hotel            0         13              2015   
4       Resort Hotel            0         14              2015   
...              ...          ...        ...               ...   
119385    City Hotel            0         23              2017   
119386    City Hotel            0        102              2017   
119387    City Hotel            0         34              2017   
119388    City Hotel            0        109              2017   
119389    City Hotel            0        205              2017   

       arrival_date_month  arrival_date_week_number arrival_date_day_of_month  \
0                    July                        27                         1   
1                    July                        27                         1   
2                    July                        27                         1   
3                    July                        27                         1   
4                    July                        27                         1   
...                   ...                       ...                       ...   
119385             August                        35                        30   
119386             August                        35                        31   
119387             August                        35                        31   
119388             August                        35                        31   
119389             August                        35                        29   

        stays_in_weekend_nights  stays_in_week_nights  adults  ...  \
0                             0                     0       2  ...   
1                             0                     0       2  ...   
2                             0                     1       1  ...   
3                             0                     1       1  ...   
4                             0                     2       2  ...   
...                         ...                   ...     ...  ...   
119385                        2                     5       2  ...   
119386                        2                     5       3  ...   
119387                        2                     5       2  ...   
119388                        2                     5       2  ...   
119389                        2                     7       2  ...   

        reservation_status  reservation_status_date                name  \
0                Check-Out               2015-07-01       Ernest Barnes   
1                Check-Out               2015-07-01        Andrea Baker   
2                Check-Out               2015-07-02      Rebecca Parker   
3                Check-Out               2015-07-02        Laura Murray   
4                Check-Out               2015-07-03         Linda Hines   
...                    ...                      ...                 ...   
119385           Check-Out               2017-09-06     Claudia Johnson   
119386           Check-Out               2017-09-07      Wesley Aguilar   
119387           Check-Out               2017-09-07        Mary Morales   
119388           Check-Out               2017-09-07  Caroline Conley MD   
119389           Check-Out               2017-09-07      Ariana Michael   

                              email  phone-number       credit_card  \
0       Ernest.Barnes31@outlook.com  669-792-1661  ************4322   
1            Andrea_Baker94@aol.com  858-637-6955  ************9157   
2        Rebecca_Parker@comcast.net  652-885-2745  ************3734   
3                 Laura_M@gmail.com  364-656-8427  ************5677   
4                LHines@verizon.com  713-226-5883  ************5498   
...                             ...           ...               ...   
119385    

## adults

In [8]:
df.adults.describe()

count    119390.000000
mean          1.856403
std           0.579261
min           0.000000
25%           2.000000
50%           2.000000
75%           2.000000
max          55.000000
Name: adults, dtype: float64

**Kolom adults memiliki nilai minimal 0, ini sebuah kesalahan dimana booking harus ada orang dewasa**

In [9]:
df[df['adults']<1].shape

(403, 37)

Ada 403 baris yang memiliki kolom adults dibawah 1

Drop baris dengan nilai dibawah 1

In [10]:
df_clean = df[df['adults']>0]
df_clean.shape

(118987, 37)

## arrival_date_week_number

In [11]:
df.arrival_date_week_number.unique()

array([27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43,
       44, 45, 46, 47, 48, 49, 50, 51, 52, 53,  1,  2,  3,  4,  5,  6,  7,
        8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24,
       25, 26], dtype=int64)

Kolom `arrival_date_week_number` akan dihapus karena tidak berulang


In [12]:
df_clean = df_clean.drop('arrival_date_week_number', axis=1)

## is_repeated_guest

In [13]:
df.is_repeated_guest.value_counts()

0    115580
1      3810
Name: is_repeated_guest, dtype: int64

Data terlihat imbalance

## previous_cancellations

In [14]:
df.previous_cancellations.value_counts()

0     112906
1       6051
2        116
3         65
24        48
11        35
4         31
26        26
25        25
6         22
19        19
5         19
14        14
13        12
21         1
Name: previous_cancellations, dtype: int64

Akan diubah valuenya:
- 0 > tidak pernah cancel
- 1 > pernah cancel 1x atau lebih

In [15]:
df_clean['previous_cancellations'] = df_clean['previous_cancellations'].apply(lambda x: 0 if x == 0 else 1)
df_clean['previous_cancellations'].value_counts()

0    112510
1      6477
Name: previous_cancellations, dtype: int64

## previous_bookings_not_canceled

In [16]:
df.previous_bookings_not_canceled.value_counts()

0     115770
1       1542
2        580
3        333
4        229
       ...  
47         1
49         1
50         1
51         1
72         1
Name: previous_bookings_not_canceled, Length: 73, dtype: int64

Akan dihapus karena redundant dengan `previous_cancellations`

In [17]:
df_clean = df_clean.drop('previous_bookings_not_canceled', axis=1)

## Kolom `reserved_room_type` dan `assigned_room_type`

Buat kolom `room_as_ordered` dari perbandingan kolom `reserved_room_type` dan `assigned_room_type`

In [18]:
df_clean['room_as_ordered'] = np.where(df_clean['reserved_room_type'] == df_clean['assigned_room_type'], 1, 0)

## agent

Berisi id dari agent travel, akan dihapus

In [19]:
df_clean = df_clean.drop('agent', axis=1)

## company

Berisi id dari perusahaan, akan dihapus

In [20]:
df_clean = df_clean.drop('company', axis=1)

## reservation_status

Redundant dengan kolom `is_canceled`, akan dihapus

In [21]:
df_clean = df_clean.drop('reservation_status', axis=1)

## Kolom `reservation_status_date`

Tidak terpakai

In [22]:
df_clean = df_clean.drop('reservation_status_date', axis=1)

## Kolom `name`

Dihapus saja

In [23]:
df_clean = df_clean.drop('name', axis=1)

## Kolom `email`

Dihapus saja

In [24]:
df_clean = df_clean.drop('email', axis=1)

## Kolom `phone-number`

Dihapus saja

In [25]:
df_clean = df_clean.drop('phone-number', axis=1)

## Kolom `credit_card`

Dihapus saja

In [26]:
df_clean = df_clean.drop('credit_card', axis=1)